File: kotlin/src/main/kotlin/ai/ensoul/sdk/EnsoulClient.kt
package ai.ensoul.sdk

import ai.ensoul.sdk.resources.*
import java.io.Closeable

class EnsoulClient private constructor(
    private val config: ClientConfig,
    internal val httpClient: EnsoulHttpClient,
) : Closeable {

    val personas: Personas = Personas(httpClient)
    val chat: Chat = Chat(httpClient)
    val domains: Domains = Domains(httpClient)
    val simulations: Simulations = Simulations(httpClient)
    val aggregate: Aggregate = Aggregate(httpClient)
    val memory: Memory = Memory(httpClient)
    val sessions: Sessions = Sessions(httpClient)
    val frameworks: Frameworks = Frameworks(httpClient)
    val auth: AuthResource = AuthResource(httpClient)
    val health: Health = Health(httpClient)
    val info: Info = Info(httpClient)

    override fun close() {
        httpClient.close()
    }

    companion object {
        const val VERSION = "0.1.0"

        operator fun invoke(
            apiKey: String? = null,
            baseUrl: String? = null,
            bearerToken: String? = null,
            timeout: Long = DEFAULT_TIMEOUT,
            maxRetries: Int = DEFAULT_MAX_RETRIES,
            customHeaders: Map<String, String> = emptyMap(),
        ): EnsoulClient {
            val resolvedApiKey = apiKey ?: System.getenv("ENSOUL_API_KEY")
            val resolvedBaseUrl = baseUrl ?: System.getenv("ENSOUL_BASE_URL") ?: DEFAULT_BASE_URL
            val config = ClientConfig(
                baseUrl = resolvedBaseUrl,
                apiKey = resolvedApiKey,
                bearerToken = bearerToken,
                timeout = timeout,
                maxRetries = maxRetries,
                customHeaders = customHeaders,
            )
            return EnsoulClient(config, EnsoulHttpClient(config))
        }

        internal fun withHttpClient(
            config: ClientConfig,
            httpClient: EnsoulHttpClient,
        ): EnsoulClient = EnsoulClient(config, httpClient)
    }
}

File: unity/Runtime/EnsoulClient.cs
using System;
using Ensoul.Resources;

namespace Ensoul
{
    public class EnsoulClient : IDisposable
    {
        public const string Version = "0.1.0";

        public PersonasResource Personas { get; }
        public ChatResource Chat { get; }
        public DomainsResource Domains { get; }
        public SimulationsResource Simulations { get; }
        public AggregateResource Aggregate { get; }
        public MemoryResource Memory { get; }
        public SessionsResource Sessions { get; }
        public FrameworksResource Frameworks { get; }
        public AuthResource Auth { get; }
        public HealthResource Health { get; }
        public InfoResource Info { get; }
        public AuditResource Audit { get; }

        private readonly EnsoulHttpClient _http;

        public EnsoulClient(EnsoulConfig config)
        {
            _http = new EnsoulHttpClient(config);
            Personas    = new PersonasResource(_http);
            Chat        = new ChatResource(_http);
            Domains     = new DomainsResource(_http);
            Simulations = new SimulationsResource(_http);
            Aggregate   = new AggregateResource(_http);
            Memory      = new MemoryResource(_http);
            Sessions    = new SessionsResource(_http);
            Frameworks  = new FrameworksResource(_http);
            Auth        = new AuthResource(_http);
            Health      = new HealthResource(_http);
            Info        = new InfoResource(_http);
            Audit       = new AuditResource(_http);
        }

        public EnsoulClient(string apiKey)
            : this(new EnsoulConfig(apiKey: apiKey)) { }

        /// Create a client using a custom HttpMessageHandler (for testing).
        public static EnsoulClient WithHttpClient(EnsoulConfig config, System.Net.Http.HttpMessageHandler handler)
        {
            var http = new EnsoulHttpClient(config, handler);
            return new EnsoulClient(http);
        }

        private EnsoulClient(EnsoulHttpClient http)
        {
            _http       = http;
            Personas    = new PersonasResource(_http);
            Chat        = new ChatResource(_http);
            Domains     = new DomainsResource(_http);
            Simulations = new SimulationsResource(_http);
            Aggregate   = new AggregateResource(_http);
            Memory      = new MemoryResource(_http);
            Sessions    = new SessionsResource(_http);
            Frameworks  = new FrameworksResource(_http);
            Auth        = new AuthResource(_http);
            Health      = new HealthResource(_http);
            Info        = new InfoResource(_http);
            Audit       = new AuditResource(_http);
        }

        public void Dispose() => _http.Dispose();
    }
}
